presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionMatcher.java

```diff
@@ -0,0 +1,294 @@
+        }
+
+        @Override
+        public Boolean visitConstant(ConstantExpression actual, RowExpression expected)
+        {
+            if (!(expected instanceof ConstantExpression)) {
+                return false;
+            }
+
+            return constantExpressionEquals(actual, (ConstantExpression) expected, aliases);
+        }
+
+        @Override
+        public Boolean visitLambda(LambdaDefinitionExpression actual, RowExpression expected)
+        {
+            if (!(expected instanceof LambdaDefinitionExpression)) {
+                return false;
+            }
+
+            return lambdaDefinitionExpressionEquals(actual, (LambdaDefinitionExpression) expected);
+        }
+
+        @Override
+        public Boolean visitVariableReference(VariableReferenceExpression reference, RowExpression expected)
+        {
+            if (!(expected instanceof VariableReferenceExpression)) {
+                return false;
+            }
+
+            return variableReferenceExpressionEquals(reference, (VariableReferenceExpression) expected);
+        }
+
+        @Override
+        public Boolean visitSpecialForm(SpecialFormExpression specialForm, RowExpression expected)
+        {
+            if (!(expected instanceof SpecialFormExpression)) {
+                return false;
+            }
+            return specialFormExpressionEquals(specialForm, (SpecialFormExpression) expected);
+        }
+    }
+
+    private boolean specialFormExpressionEquals(SpecialFormExpression actual, SpecialFormExpression expected)
+    {
+        if (!actual.getType().equals(expected.getType())) {
+            return false;
+        }
+
+        if (!actual.getForm().equals(expected.getType())) {
+            return false;
+        }
+
+        return isSame(actual.getArguments(), expected.getArguments());
+    }
+
+    private boolean isSame(List<RowExpression> actual, List<RowExpression> expected)
+    {
+        if (actual.size() != expected.size()) {
+            return false;
+        }
+
+        Iterator<RowExpression> actualArgsItr = actual.iterator();
+        Iterator<RowExpression> expectedArgsItr = expected.iterator();
+
+        while (actualArgsItr.hasNext() && expectedArgsItr.hasNext()) {
+            RowExpression actualArgument = actualArgsItr.next();
+            RowExpression expectedArgument = expectedArgsItr.next();
+
+            boolean isMatch = match(actualArgument, expectedArgument);
+            if (!isMatch) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean inputReferenceExpressionEquals(InputReferenceExpression actual, InputReferenceExpression expected)
+    {
+        if (!actual.getType().equals(expected.getType())) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean variableReferenceExpressionEquals(VariableReferenceExpression actual, VariableReferenceExpression expected)
```

Comment:
Make it protected so that we can override it for new behaviors. Also, we might want to pass in `aliases` to compare aliases.